server/gfmd/user.c

```diff
@@ -419,6 +355,154 @@ user_enter_auth_id(
 	return (GFARM_ERR_NO_ERROR);
 }
 
+static gfarm_error_t
+user_auth_id_modify_internal(struct user *u,
+		char *auth_id_type, char *auth_user_id,
+		bool *need_to_update_dbp, bool *need_to_addp)
+{
+	enum auth_user_id_type auth_user_id_type = 0;
+	char *old_auth_user_id = NULL;
+	struct gfarm_hash_entry *entry;
+	gfarm_error_t e;
+
+	*need_to_update_dbp = false;
+
+	if ((e = gfarm_auth_user_id_type_from_name(
+		(char *)auth_id_type,
+		&auth_user_id_type)) != GFARM_ERR_NO_ERROR) {
+		return (e);
+	}
+
+	old_auth_user_id = u->auth_user_id[auth_user_id_type];
+
+	if (old_auth_user_id != NULL) {
+		if (strcmp(u->auth_user_id[auth_user_id_type],
+				auth_user_id) == 0) {
+			return (GFARM_ERR_NO_ERROR);
+		}
+
+		struct user_auth_key key = {
+			auth_user_id_type,
+			u->auth_user_id[auth_user_id_type]
+		};
+
+		entry = gfarm_hash_lookup(user_id_map, &key, sizeof(key));
+		if (entry != NULL) {
+			gfarm_hash_purge(user_id_map, &key, sizeof(key));
+		} else {
+			return (GFARM_ERR_ALREADY_EXISTS);
+		}
+		*need_to_addp = false;
+	} else {
+		*need_to_addp = true;
+	}
+
+	u->auth_user_id[auth_user_id_type] = strdup_log(auth_user_id,
+						"user_auth_id_modify_internal");
+
+	if (u->auth_user_id[auth_user_id_type] == NULL)
+		return (GFARM_ERR_NO_MEMORY);
+
+	e = user_enter_auth_id(auth_user_id_type,
+		       u->auth_user_id[auth_user_id_type], u);
+
+	if (e != GFARM_ERR_NO_ERROR) {
+		free(u->auth_user_id[auth_user_id_type]);
+		u->auth_user_id[auth_user_id_type] = old_auth_user_id;
+	} else {
+		free(old_auth_user_id);
+	}
+
+	*need_to_update_dbp = true;
+
+	return (e);
+}
+
+gfarm_error_t
+user_auth_id_modify(struct user *user, char *auth_id_type,
+	char *auth_user_id)
+{
+	gfarm_error_t e;
+	bool need_to_update_db;
+	bool need_to_add;
+
+	e = user_auth_id_modify_internal(user,
+		auth_id_type, auth_user_id,
+		&need_to_update_db,
+		&need_to_add);
+
+	if (e == GFARM_ERR_INVALID_ARGUMENT) {
+		gflog_fatal(GFARM_MSG_UNFIXED,
+			"user %s: unknown auth_id_type %s, \
+			slave gfmd is older than gfmd",
+			user->ui.username,
+			auth_id_type);
+	} else if (e == GFARM_ERR_ALREADY_EXISTS) {
+		gflog_fatal(GFARM_MSG_UNFIXED,
+			"user %s: unexpected inconsistency, \
+			adding duplicate auth_user_id_type:%s, \
+			auth_user_id:%s",
+			user->ui.username,
+			auth_id_type,
+			auth_user_id);
+	}
+
+	return (e);
+
+}
+
+static gfarm_error_t
+user_auth_id_remove_internal(struct user *user,
+	char *auth_id_type, bool *need_to_update_dbp)
+{
+	enum auth_user_id_type auth_user_id_type = 0;
+	gfarm_error_t e;
+
+	if ((e = gfarm_auth_user_id_type_from_name(
+		(char *)auth_id_type,
+		&auth_user_id_type)) != GFARM_ERR_NO_ERROR) {
+		return (e);
+	}
+
+	if (user->auth_user_id[auth_user_id_type] == NULL) {
+		*need_to_update_dbp = false;
+		return (GFARM_ERR_NO_ERROR);
+	} else {
+		struct user_auth_key key = {
+			auth_user_id_type,
+			user->auth_user_id[auth_user_id_type]
+		};
+
+		gfarm_hash_purge(user_id_map, &key, sizeof(key));
```

Comment:
if gfarm_hash_purge() fails (i.e. if it returns 0), please call gflog_fatal().
That is unexpected internal inconsistency.